### DISCUSSION TITLE
IContentDefinitionManager dependencies

### DISCUSSION THREAD

**Discussion by alexisborn** [OrchardCMS/OrchardCore#8502] [Q&A] [answered]
Hi,

I'm playing with DataMigration. So I've added:

```
public class MyMigrations : DataMigration
{
        private readonly IContentDefinitionManager _contentDefinitionManager;

        public MyMigrations(IContentDefinitionManager contentDefinitionManager)
        {
            _contentDefinitionManager = contentDefinitionManager;
        }

        public int Create()
        {...}
```

My Startup.cs:

```
public class Startup
    {
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddIdGeneration();
            services.AddContentManagement();
            services.AddScoped<IDataMigration, MyMigrations>();
            services.AddOrchardCms();
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            app.UseRouting();
            app.UseOrchardCore();
        }
    }
```

I see the following exception:
`AggregateException: Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: OrchardCore.Environment.Cache.ICacheContextProvider Lifetime: Scoped ImplementationType: OrchardCore.ContentManagement.Cache.ContentDefinitionCacheContextProvider': Unable to resolve service for type 'OrchardCore.Environment.Cache.ISignal' while attempting to activate 'OrchardCore.ContentManagement.ContentDefinitionManager'.) (Error while validating the service descriptor 'ServiceType: OrchardCore.ContentManagement.Metadata.IContentDefinitionManager Lifetime: Scoped ImplementationType: OrchardCore.ContentManagement.ContentDefinitionManager': Unable to resolve service for type 'OrchardCore.Environment.Cache.ISignal' while attempting to activate 'OrchardCore.ContentManagement.ContentDefinitionManager'.) (Error while validating the service descriptor 'ServiceType: OrchardCore.ContentManagement.IContentDefinitionStore Lifetime: Scoped ImplementationType: OrchardCore.ContentManagement.DatabaseContentDefinitionStore': Unable to resolve service for type 'YesSql.ISession' while attempting to activate 'OrchardCore.ContentManagement.DatabaseContentDefinitionStore'.) (Error while validating the service descriptor 'ServiceType: OrchardCore.ContentManagement.IContentManager Lifetime: Scoped ImplementationType: OrchardCore.ContentManagement.DefaultContentManager': Unable to resolve service for type 'OrchardCore.Environment.Cache.ISignal' while attempting to activate 'OrchardCore.ContentManagement.ContentDefinitionManager'.) (Error while validating the service descriptor 'ServiceType: OrchardCore.Data.Migration.IDataMigration Lifetime: Scoped ImplementationType: OrchardCore.ContentManagement.Records.Migrations': Unable to resolve service for type 'OrchardCore.Environment.Cache.ISignal' while attempting to activate 'OrchardCore.ContentManagement.ContentDefinitionManager'.) (Error while validating the service descriptor 'ServiceType: OrchardCore.ContentManagement.Handlers.IContentHandler Lifetime: Scoped ImplementationType: OrchardCore.ContentManagement.Handlers.ContentPartHandlerCoordinator': Unable to resolve service for type 'OrchardCore.Environment.Cache.ISignal' while attempting to activate 'OrchardCore.ContentManagement.ContentDefinitionManager'.) (Error while validating the service descriptor 'ServiceType: OrchardCore.Data.Migration.IDataMigration Lifetime: Scoped ImplementationType: TestOrchard.MyMigrations': Unable to resolve service for type 'OrchardCore.Environment.Cache.ISignal' while attempting to activate 'OrchardCore.ContentManagement.ContentDefinitionManager'.)`

Let me know please, what should I add?

@deanmarcussen @jtkech Any help would be much appreciated!

Thanks!

**Comment by alexisborn** [comment, marked as answer]
could anybody please help with it?

**Reply by jtkech** [reply]
@alexisborn 

        public void ConfigureServices(IServiceCollection services)
        {
            // Registered for you through 'AddOrchardCms()' at the tenant level (not the app level)
            // services.AddIdGeneration();
            // services.AddContentManagement();

            // Intended to be registered at the tenant level, e.g. through a module startup
            // services.AddScoped<IDataMigration, MyMigrations>();

            // But the app can behaves as a module by using our helpers on the OrchardcoreBuilder
            // returned by 'AddOrchardCms()'
            services.AddOrchardCms()
                .ConfigureServices(tenantServices =>
                {
                    // As it is done through the `ConfigureServices()` of a module startup
                    tenantServices.AddScoped<IDataMigration, MyMigrations>();
                });
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            // Will be done for you at the tenant level (not app level) when building the tenant / shell pipeline
            // app.UseRouting();

            app.UseOrchardCore();
        }
    }

**Reply by alexisborn** [reply]
@jtkech  Thank you for the answer!

Now my Startup.cs looks as follows:

```
 public void ConfigureServices(IServiceCollection services)
    {
        services.AddOrchardCms()
            .ConfigureServices(tenantServices =>
            {
                tenantServices.AddScoped<IDataMigration, MyMigrations>();
            });
    }

    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
    {
        app.UseOrchardCore();
    }
}
```

The setup page is shown, after filling all the required fields in and clicking on the "Finish Setup" button I see the following:

`An unhandled exception occurred while processing the request.
InvalidOperationException: Unable to resolve service for type 'OrchardCore.ContentManagement.Metadata.IContentDefinitionManager' while attempting to activate 'TestOrchard.MyMigrations'.
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.CreateArgumentCallSites(Type serviceType, Type implementationType, CallSiteChain callSiteChain, ParameterInfo[] parameters, bool throwIfCallSiteNotFound)`

**Reply by jtkech** [reply]
@alexisborn 

Hmm, yes, when we activate a given shell / tenant, we execute the migrations steps (if needed) of the modules that are enabled, during a setup at some point we create a temporary and light shell / tenant container but that doesn't include any module / feature yet, so when we activate it no migrations is run, but the "app's module" is always enabled, so your migration run, but because the `OC.Contents` module is not enabled yet, the `IContentDefinitionManager` is not registered.

So, you would need to create a module to be able to register and use your migrations

Hmm, what you can try is to inject `IServiceProvider`, then resolve lazily `_services.GetService<IContentDefinitionManager>()`, not `GetRequiredService<>()`, and then do nothing if it is null.

**Reply by alexisborn** [reply]
Hello @jtkech 

Thanks a lot for your patience!

Both options are working, but if I use the one with IServiceProvider, the ContentType options (Draftable, Versionable and so on) are false by default (If I do the module approach they are true):
```

private readonly IContentDefinitionManager _contentDefinitionManager;

        public MyMigrations(IServiceProvider serviceProvider)
        {
            _contentDefinitionManager = serviceProvider.GetService<IContentDefinitionManager>();
        }

        public int Create()
        {
            if (_contentDefinitionManager == null)
            {
                return 0;
            }
            _contentDefinitionManager.AlterTypeDefinition("HomePage", type => type
                    .Draftable()
                    .Versionable()
                    .Listable()
                    .Creatable()
                    .Securable()
                    .WithPart(nameof(TitlePart), part => part
                    .WithPosition("1"))
                    .WithPart(nameof(TestPart), part => part
                    .WithPosition("2"))
                );

            _contentDefinitionManager.AlterPartDefinition(nameof(TestPart), part => part
                .WithField("Image", field => field
                    .OfType(nameof(MediaField))
                    .WithDisplayName("Image")
                )
                .WithField("Price", field => field
                    .OfType(nameof(NumericField))
                    .WithDisplayName("Price")
                )
            );

            return 1;
        }
```
Is this expectable?

One more question: If I need to add more fields/parts/types (or remove something) later should I create one more migration (TestOrchard.MySecondMigrations) and register it or is it possible to use somehow just one (MyMigrations)?

```
services.AddScoped<IDataMigration, MyMigrations>();
services.AddScoped<IDataMigration, MySecondMigrations>();
```

**Reply by jtkech** [reply]
@alexisborn 

> Both options are working, but if I use the one with IServiceProvider, the ContentType options (Draftable, Versionable and so on) are false by default (If I do the module approach they are true):

Maybe it is not executed in the right order because it doesn't depend on the `OC.ContentTypes` module / feature, with a module you can add this dependency in the module `Manifet.cs` file (see in our manifest files).

So maybe better to use a module

But the `ConfigureSevices` has an optional `order` parameter equal to 0 by default, so try an higher value, e.g.

        services.AddOrchardCms()
            .ConfigureServices(tenantServices =>
            {
                tenantServices.AddScoped<IDataMigration, MyMigrations>();
            }, order: 100);

> One more question: If I need to add more fields/parts/types (or remove something) later should I create one more migration (TestOrchard.MySecondMigrations) and register it or is it possible to use somehow just one (MyMigrations)?

You don't need another migrations class, in the existing one you need to add an `UpgradeFrom1()` method migration step that upgrade from version 1 to version 2 (from n to n + 1), and update your `Create()` to always return the last version.

        public int Create()
        {
            // ...

            // Shortcut other migration steps on new content definition schemas.
            return 2;
        }

        // This code can be removed in a later version.
        public int UpdateFrom1()
        {
            // migration step

            return 2;
        }

**Reply by jtkech** [reply]
Okay cool ;)

Mark it as answered if do not have other question

**Reply by alexisborn** [reply]
just checked it out: adding the "order: 100" resolved the problem, using UpdateFrom{n} works as well

@jtkech Thank you very much!
